src/mpi/coll/allreduce/allreduce_intra_recursive_multiplying.c

```diff
@@ -0,0 +1,233 @@
+ * Setting k according to the network hierarchy (e.g., the number of NICs
+ * in a node) can improve performance.
+ */
+
+
+int MPIR_Allreduce_intra_recursive_multiplying(const void *sendbuf,
+                                            void *recvbuf,
+                                            MPI_Aint count,
+                                            MPI_Datatype datatype,
+                                            MPI_Op op,
+                                            MPIR_Comm * comm_ptr, 
+                                            const int k,
+                                            MPIR_Errflag_t errflag)
+{
+    /* If the op is non-commutative, fall back to recursive doubling */
+    if(!MPIR_Op_is_commutative(op)) {
+      return MPIR_Allreduce_intra_recursive_doubling(sendbuf, recvbuf, count, datatype,
+                                                op, comm_ptr, errflag);
+    }
+
+    /*
+     * If the comm size is not a multiple of k, fall back to recursive
+     * doubling. Recursive multiplying performs best when k is a multiple
+     * of comm size, so this case should not be used anyways.
+     */
+    int comm_size, rank, virt_rank;
+    comm_size = comm_ptr->local_size;
+    rank = comm_ptr->rank;
+    if(comm_size % k) {
+      return MPIR_Allreduce_intra_recursive_doubling(sendbuf, recvbuf, count, datatype,
+                                                op, comm_ptr, errflag);
+    }
+
+    MPIR_CHKLMEM_DECL(2);
+    int mpi_errno = MPI_SUCCESS;
+    int dst, pofk;
+    int comm_size_div_pofk, comm_size_mod_pofk, group, group_size, group_leader;
+    int radix_size, next_radix_size, next_radix_root, radix_loc;
+    int dst_root, dst_root_send, dst_group_size, real_dst;
+    MPI_Aint true_extent, true_lb, extent;
+    void *tmp_buf, *red_buf;
+    MPIR_Request **reqs;
+    int i, p;
+    int num_reqs = 0, parity = 0;
+
+    /* need to allocate temporary buffer to store incoming data */
+    MPIR_Type_get_true_extent_impl(datatype, &true_lb, &true_extent);
+    MPIR_Datatype_get_extent_macro(datatype, extent);
+    extent = MPL_MAX(extent, true_extent);
+
+    MPIR_CHKLMEM_MALLOC(tmp_buf, void *, (k - 1) * count * extent, mpi_errno,
+                        "temporary buffer", MPL_MEM_BUFFER); // cut to half
+    MPIR_CHKLMEM_MALLOC(reqs, MPIR_Request **, (2 * (k - 1) * sizeof(MPIR_Request *)), mpi_errno,
+                        "reqs", MPL_MEM_BUFFER);
+
+    /* adjust for potential negative lower bound in datatype */
+    tmp_buf = (void *) ((char *) tmp_buf - true_lb);
+
+    /* copy local data into recvbuf */
+    if (sendbuf != MPI_IN_PLACE) {
+        mpi_errno = MPIR_Localcopy(sendbuf, count, datatype, recvbuf, count, datatype);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+    /* get nearest power-of-two less than or equal to comm_size */
+    p = (int) (log(comm_size) / log(k));
+    pofk = (int) lround(pow(k, p));
+
+    /* Compute group-specific information */
```

Comment:
The comment should explain your strategy. At this point, I have no clue what you are trying to do or what is a `group`.

Reply:
I updated the top comment in the file (starts at line 13) to be much more descriptive about the algorithm. Let me know if there is anything the remains unclear about the overall strategy.